### DISCUSSION TITLE
s3 presigned request does not work with custom endpoint and no region set

### DISCUSSION THREAD

**Discussion by ibotty** [smithy-lang/smithy-rs#3404] [Q&A]
If using an aws-config with custom endpoint, but no region the presigned request is not complete.  If setting the region it works allright.

Is that expected?  If so it would be great to at least document it or better statically prohibit using it without region.

```
let aws_config = aws_config::from_env()
    .endpoint_url(...)
    //.region("us-east-1")
    .load()
    .await;
let client = aws_sdk_s3::Client::from_conf(s3_config);
let req = client.get_object()
    .bucket(..)
    .key("test-key")
    .presigned(..)
    .await.unwrap();
assert_eq(req.uri, "/test-key/?x-id=GetObject")
```

**Comment by rcoh** [comment]
signing requires a region as an inherent part of signing, so any request needs a region. Unfortunately, there isn't a practical way to prohibit this statically, since, e.g. a region could be set as late as config override when customizing the request.

**Reply by ibotty** [reply]
Thank you for your answer.

If a custom endpoint is set, does it matter what the region is set to?  For my self-hosted s3 stores (Ceph RadosGW and Minio) it did not matter which region I set.

Is it possible to set a default region if none is set and a custom endpoint is used?

**Reply by rcoh** [reply]
does the request still need to be signed? You could set a default region on your client. We need to know _something_ to put in the "credential scope" section of the signature, you can't sign without it.

**Reply by ibotty** [reply]
Yes, it has to be signed.  I do now set a default region, but I do think that's not the best API.  My testing shows the region does not matter when using a custom endpoint, so it should be optional.  Am I wrong in this regard?

**Reply by rcoh** [reply]
If the SDK was designed to exclusively work with services like Ceph, that may be a reasonable strategy. However, for most AWS customers, there is no _good_ default region (and setting one automatically or having a default) would almost certainly be the wrong thing.

**Reply by ibotty** [reply]
I do understand that (and don't want to convince you otherwise), but do users on AWS set an endpoint URL?  I don't when deploying on AWS, but I am most likely missing things.

**Reply by rcoh** [reply]
There are a few use cases where it is required, things like S3 VPC endpoints, localstack, etc. It's fairly rare, but does happen.

**Reply by ibotty** [reply]
Ah, I see.  Maybe it could be at least documented, that setting an endpoint is not sufficient without also setting a region somehow.

Thank you for the friendly discussion.  Very appreciated.
